cmd/node_agent/na/nodeagent.go

```diff
@@ -78,46 +42,106 @@ type platformSpecificRequest interface {
+	if cfg.IstioCAAddress == "" {
+		return nil, fmt.Errorf("Istio CA address is empty")
+	}
+	dialOptions, optionErr := pr.GetDialOptions(cfg)
+	if optionErr != nil {
+		return nil, optionErr
+	}
+	return &cAGrpcClientImpl{
+		&cfg.IstioCAAddress,
+		dialOptions,
+	}, nil
+}
+
+// cAGrpcClientImpl is a implementation of GRPC client to talk to CA.
+type cAGrpcClientImpl struct {
+	cAAddress   *string
+	dialOptions []grpc.DialOption
+}
+
+// SendCSR sends CSR to CA through GRPC.
+func (c *cAGrpcClientImpl) SendCSR(req *pb.Request) (*pb.Response, error) {
+	conn, err := grpc.Dial(*c.cAAddress, c.dialOptions...)
+	if err != nil {
+		glog.Errorf("Failed to dial %s: %s", *c.cAAddress, err)
+		return nil, err
+	}
+	defer func() {
+		if closeErr := conn.Close(); closeErr != nil {
+			glog.Errorf("Failed to close connection")
+		}
+	}()
+	client := pb.NewIstioCAServiceClient(conn)
+	resp, err := client.HandleCSR(context.Background(), req)
+	if err != nil {
+		glog.Errorf("CSR request failed %v", err)
+		return nil, err
+	}
+	return resp, nil
+}
+
 // The real node agent implementation. This implements the "Start" function
 // in the NodeAgent interface.
 type nodeAgentInternal struct {
 	// Configuration specific to Node Agent
-	config *Config
-	pr     platformSpecificRequest
+	config   *Config
+	pr       platformSpecificRequest
+	cAClient CAGrpcClient
 }
 
 // Start the node Agent.
-func (na *nodeAgentInternal) Start() {
-
+func (na *nodeAgentInternal) Start() error {
 	if na.config == nil {
-		glog.Fatalf("Node Agent configuration is nil")
+		retErr := fmt.Errorf("node Agent configuration is nil")
+		glog.Error(retErr)
+		return retErr
 	}
 
 	if !na.pr.IsProperPlatform() {
-		glog.Fatalf("Node Agent is not running on the right platform")
+		retErr := fmt.Errorf("node Agent is not running on the right platform")
+		glog.Error(retErr)
+		return retErr
 	}
 
 	glog.Infof("Node Agent starts successfully.")
+
 	retries := 0
-	retrialInterval := certRequestRetrialInterval
+	retrialInterval := na.config.CSRInitialRetrialInterval
 	success := false
 	for {
+		privKey, req, reqErr := na.createRequest()
+		if reqErr != nil {
+			return reqErr
+		}
+
 		glog.Infof("Sending CSR (retrial #%d) ...", retries)
-		privKey, resp, err := na.sendCSR()
+
+		resp, err := na.cAClient.SendCSR(req)
 		if err == nil && resp != nil && resp.IsApproved {
-			cert, certErr := pki.ParsePemEncodedCertificate(resp.SignedCertChain)
-			if certErr != nil {
-				glog.Errorf("Error getting TTL from approved cert: %v", certErr)
+			waitTime, ttlErr := na.getWaitTimeFromCert(
+				resp.SignedCertChain, time.Now(), na.config.CSRGracePeriodPercentage)
+			if ttlErr != nil {
+				glog.Errorf("Error getting TTL from approved cert: %v", ttlErr)
 				success = false
 			} else {
-				certTTL := cert.NotAfter.Sub(cert.NotBefore)
-				// Wait until the grace period starts.
-				waitTime := certTTL - time.Duration(certRenewalGracePeriodPercentage/100)*certTTL
 				timer := time.NewTimer(waitTime)
```

Comment:
nit: move this closer to where it's used (line 145)

Reply:
Done.